crates/core/src/host/instance_env.rs

```diff
@@ -326,56 +328,54 @@ impl InstanceEnv {
 
     #[tracing::instrument(skip_all)]
     pub fn iter(&self, table_id: u32) -> impl Iterator<Item = Result<Vec<u8>, NodesError>> {
-        use genawaiter::{sync::gen, yield_, GeneratorState};
+        use genawaiter::{
+            sync::{gen, Gen},
+            yield_,
+        };
 
         // Cheap Arc clones to untie the returned iterator from our own lifetime.
         let relational_db = self.dbic.relational_db.clone();
         let tx = self.tx.clone();
 
-        // For now, just send buffers over a certain fixed size.
-        fn should_yield_buf(buf: &Vec<u8>) -> bool {
-            const SIZE: usize = 64 * 1024;
-            buf.len() >= SIZE
-        }
-
-        let mut generator = Some(gen!({
+        // TODO(jgilles): replace this with a proper iterator struct for the mild perf gain
+        // (you will need to use Pin)
+        let generator = gen!({
             let stdb = &*relational_db;
-            let tx = &mut *tx.get()?;
+            let mut tx = &mut *tx.get()?;
 
-            let mut buf = Vec::new();
             let schema = stdb.row_schema_for_table(tx, table_id)?;
-            schema.encode(&mut buf);
-            yield_!(buf);
+            yield_!(IterItem::Header(schema));
 
-            let mut buf = Vec::new();
             for row in stdb.iter(tx, table_id)? {
-                if should_yield_buf(&buf) {
-                    yield_!(buf);
-                    buf = Vec::new();
-                }
-                row.view().encode(&mut buf);
-            }
-            if !buf.is_empty() {
-                yield_!(buf)
+                yield_!(IterItem::Row(row.view().clone()));
             }
 
             Ok(())
-        }));
-
-        std::iter::from_fn(move || match generator.as_mut()?.resume() {
-            GeneratorState::Yielded(bytes) => Some(Ok(bytes)),
-            GeneratorState::Complete(res) => {
-                generator = None;
-                match res {
-                    Ok(()) => None,
-                    Err(err) => Some(Err(err)),
+        });
+
+        struct AsIterator<F: Future<Output = Result<(), NodesError>>>(Gen<IterItem, (), F>);
```

Comment:
Does this actually help in benchmarks? `std::iter::from_fn` is a zero-cost wrapper (`Fn` with captures is the same as writing struct by hand) and a bit cleaner to read than manual iterator implementation.